Germ Selection Gate Penalty

coreyostrove opened the pull request:
We recently ran into a scenario where it was desirable to perform germ selection in a manner which minimized the number of applications of a particularly expensive gate operation. To support that sort of use case this PR adds a new `algorithm_kwarg` option called `gate_penalty` for `find_germs` which is compatible with the 'greedy' and 'grasp' search algorithms which allows a user to specify and additional penalty factors for the number of instances of particular gates in a candidate germ set. Also included are some additional unit tests for both this new gate penalty as well as tests for the existing `op_penalty` option (which penalized the total number of gate operations overall).

rileyjmurray commented on pygsti/algorithms/germselection.py:
```diff
@@ -432,7 +436,7 @@ def compute_germ_set_score(germs, target_model=None, neighborhood=None,
                            randomization_strength=1e-2, score_func='all',
                            op_penalty=0.0, l1_penalty=0.0, num_nongauge_params=None,
-                           float_type=_np.cdouble):
+                           float_type=_np.cdouble, gate_penalty=None):
```
Type annotation please :)

rileyjmurray commented on test/unit/algorithms/test_germselection.py:
```diff
@@ -374,8 +374,49 @@ def test_force_germs_outside_candidate_set(self):
+                                                 assume_real=True, float_type=np.double, mode='all-Jac', algorithm='grasp',
+                                                 algorithm_kwargs={'seed':1234, 'iterations':1})
+
+        assert count_gate(germs_gate_penalty_grasp, 'Gxpi2') < count_gate(germs_default_grasp, 'Gxpi2')
```
The keyword arguments here are hard to read. I suggest something like the following.
```python
class GermSelectionPenaltyTester(GermSelectionData, BaseCase):

    common_kwargs  : dict[str, Any] = dict(
        randomize=True, seed=1234, candidate_germ_counts={7:'all upto'},
        assume_real=True, float_type=np.double, mode='compactEVD', algorithm='greedy'
    )

    def setUp(self):
        super(GermSelectionData, self).setUp()

    def test_op_penalty_greedy(self):
        germs_no_penalty_cevd = germsel.find_germs(self.target_model, **self.common_kwargs)
        germs_penalty_cevd    = germsel.find_germs(self.target_model, **self.common_kwargs, algorithm_kwargs={'op_penalty':.1})        
        assert count_ops(germs_no_penalty_cevd) > count_ops(germs_penalty_cevd)


    def test_gate_penalty_greedy(self):
        kwargs = self.common_kwargs.copy()
        kwargs.pop('mode')
        germs_no_penalty_alljac = germsel.find_germs(self.target_model, **kwargs, mode='all-Jac')     
        germs_no_penalty_cevd   = germsel.find_germs(self.target_model, **kwargs, mode='compactEVD')
        germs_penalty_alljac = germsel.find_germs(
            self.target_model, **kwargs, mode='all-Jac',    algorithm_kwargs={'gate_penalty':{'Gxpi2':.1}}
        )        
        germs_penalty_cevd = germsel.find_germs(
            self.target_model, **kwargs, mode='compactEVD', algorithm_kwargs={'gate_penalty':{'Gxpi2':.1}}
        )
        assert count_gate(germs_no_penalty_alljac, 'Gxpi2') > count_gate(germs_penalty_alljac, 'Gxpi2')
        assert count_gate(germs_no_penalty_cevd,   'Gxpi2') > count_gate(germs_penalty_cevd,   'Gxpi2')
                                           
    def test_gate_penalty_grasp(self):
        kwargs = self.common_kwargs.copy()
        kwargs['mode']      = 'all-Jac'
        kwargs['algorithm'] = 'grasp'
        germs_gate_penalty_grasp = germsel.find_germs(
            self.target_model, **kwargs, algorithm_kwargs={'seed':1234, 'iterations':1, 'gate_penalty':{'Gxpi2':.2}}
        )
        germs_default_grasp = germsel.find_germs(
            self.target_model, **kwargs, algorithm_kwargs={'seed':1234, 'iterations':1}
        )
        assert count_gate(germs_gate_penalty_grasp, 'Gxpi2') < count_gate(germs_default_grasp, 'Gxpi2')
```

rileyjmurray commented on test/unit/algorithms/test_germselection.py:
```diff
@@ -388,4 +429,17 @@ def lite_germ_selection_end_to_end_test(self):
+    num_ops = 0
+    for circuit in circuits:
+        num_ops+= circuit.num_gates
+    return num_ops
```
Missing blank line at EOF

rileyjmurray replied:
Technically it's part of the UNIX standard: https://unix.stackexchange.com/questions/18743/whats-the-point-in-adding-a-new-line-to-the-end-of-a-file.

rileyjmurray commented:
@coreyostrove, is this is an appropriate PR to look into the complex-dtype issues with compactEVD germ selection https://github.com/sandialabs/pyGSTi/issues/659? It's fine with me if you'd like to handle that separately, but I figured I'd point out the chance.

rileyjmurray commented on pygsti/algorithms/germselection.py:
```diff
@@ -648,6 +660,20 @@ def compute_composite_germ_set_score(score_fn, threshold_ac=1e6, init_n=1,
+        of the specified gate(s) in each germ. Should be specified as a dictionary whose keys are strings
+        corresponding to gate names, and whose values are the penalty factor to add for each instance of that
+        gate. E.g. {'Gcnot':2} would correspond to a penalty term where each instance of a 'Gcnot' gate
+        gets and additional 2 units added to the cost function for a candidate germ.
```
@coreyostrove and I have come to a truce. It's okay for me to use docstring injection of various kinds, but he can proceed with the normal way.

coreyostrove commented on pygsti/algorithms/germselection.py:
```diff
@@ -4256,6 +4347,18 @@ def compute_composite_germ_set_score_compactevd(current_update_cache, germ_updat
+                #alternative approach using the string 
+                #representation of the ckt.
+                num_gate_instances= germ.str.count(gate)
+                gate_score += num_gate_instances*penalty_value
```
Done

coreyostrove commented:
@rileyjmurray: I believe I've addressed all of your feedback, so this is ready for re-review. Regarding the complex data type bug, I'll tackle that in a future PR.
